### PR TITLE
Add more eslint rules

### LINKED ISSUE #959: More eslint rules
**Is your feature request related to a problem? Please describe.**
Basic eslint rules have been implemented, but there are more that would be useful to have.

**Describe the solution you'd like**
There are eslint plugins that exist to help cover cases not already covered. These will be created as linked issues with information about useable plugins

- [ ] #960
- [ ] #961
- [ ] #962
- [ ] #963
- [ ] #964

Overall, this seems to be a good resource for plugins: https://github.com/dustinspecker/awesome-eslint

### LINKED ISSUE #960: Eslint Import Rules
There are various eslint rules related to imports that it would be nice to have. Two of the main ones that come to mind are import order and two newlines after the last import. The plugin https://github.com/import-js/eslint-plugin-import looks to be a good choice for those and more.

### LINKED ISSUE #961: Eslint jsdoc
Eslint used to have some built in linting for jsdoc comments, but it no longer does (see https://eslint.org/blog/2018/11/jsdoc-end-of-life). That link mentions https://github.com/gajus/eslint-plugin-jsdoc which seems to be the best choice.

### LINKED ISSUE #962: Eslint jest
Tests should be neatly and consistently written/structured like the rest of the codebase, and eslint rules specific for jest can help with that. There seem to be two main repos for this: https://github.com/jest-community/eslint-plugin-jest and https://github.com/dangreenisrael/eslint-plugin-jest-formatting. Some research would be needed to see if both are necessary or one or the other, but from a preliminary glance it looks like both could be used.

### LINKED ISSUE #963: Eslint react
There are multiple plugins for using eslint with various react specific rules. These include ones with [accessability](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y), [react in general](https://github.com/yannickcr/eslint-plugin-react), [hooks](https://github.com/facebook/react/tree/main/packages/eslint-plugin-react-hooks), or [redux](https://github.com/DianaSuvorova/eslint-plugin-react-redux).

### LINKED ISSUE #964: Warn/error against using deprecated functions

### PR DESCRIPTION
**Pull request by TGCrystal**
This pr adds and configures the following plugins, in addition to configuring `"@typescript-eslint/naming-convention"` and a few other base eslint/typescript-eslint rules:
- https://github.com/import-js/eslint-plugin-import
- https://github.com/Simonwep/eslint-plugin-align-import
- https://github.com/jsx-eslint/eslint-plugin-jsx-a11y
- https://github.com/yannickcr/eslint-plugin-react
- https://github.com/facebook/react/tree/main/packages/eslint-plugin-react-hooks
- https://github.com/DianaSuvorova/eslint-plugin-react-redux
- https://github.com/SonarSource/eslint-plugin-sonarjs
- https://github.com/sindresorhus/eslint-plugin-unicorn
- https://github.com/jest-community/eslint-plugin-jest
- https://github.com/dangreenisrael/eslint-plugin-jest-formatting

Things left to consider:
- [ ] `@typescript-eslint/naming-convention` needs a look from @LeonMontealegre to make sure it is correct
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/destructuring-assignment.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/sort-comp.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/sort-prop-types.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/static-property-placement.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-boolean-value.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-closing-bracket-location.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-curly-brace-presence.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-curly-spacing.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-indent-props.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-newline.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-sort-props.md
- [x] https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-space-before-closing.md
- [x] https://github.com/gajus/eslint-plugin-jsdoc#check-line-alignment
- [x] https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-require-description-complete-sentence
- [x] https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/cognitive-complexity.md
- [x] https://github.com/mthadley/eslint-plugin-sort-destructure-keys

closes #960 
closes #964 
closes #963 
closes #961 
closes #959 
closes #962

### CODE REVIEW

**Comment by ml619** [comment]
Some issues I have found with changes Eslint made
src\react-app-env.d.ts: added a space after the second /
src\site\shared\containers\SelectionPopup\index.tsx line 23: changed from exporting a function to exporting a var
Eslint turns <div></div> to </div> and we probaly do not want that to happen

File: .eslintrc.js [hunk 1]
```diff
@@ -1,104 +1,78 @@
...
+        }
+    }
+
+    for (const property in object2) {
+        if (!(property in object1))
+            newObject[property] = object2[property];
+    }
+
+    return newObject;
+}
```
Comment (LeonMontealegre) [inline .eslintrc.js:51]: I believe I already have a `mergeDeep` utility in `scripts/utils`

File: .eslintrc.js [hunk 2]
```diff
@@ -1,104 +1,78 @@
...
+    imports,
+    jsdoc,
+    react,
+    jestRules,
+    custom,
+    unicorn,
+    sonarjs,
+    jsxa11y,
+]
+module.exports = mergeAll(ruleSets);
```
Comment (LeonMontealegre) [inline .eslintrc.js:78]: I would say look at how I merge things in `scripts/webpack/config/index.js`, i think the general output would be like:
```js
module.exports = mergeDeep({
    require("./linting/.core.js"),
    require("./linting/.imports.js"),
    ...
});
```

Also I think even better would be to specify all the core settings in this file, like the 
```
    "env": {
        "browser": true,
        "es6": true,
    },
    "extends": [
        "eslint:recommended",
    ],
    "ignorePatterns": ["*.js", "scripts/**"],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "project": "./tsconfig.base.json",
        "ecmaFeatures": {
            "jsx": true,
        },
    },
    "settings": {
        "react": {
            "version": "detect",
        },
    },
```

properties

Comment (TGCrystal) [reply .eslintrc.js:78]: So the whole .core file would be moved back into the main .eslintrc.js file?

File: linting/.core.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    "parserOptions": {
+        "project": "./tsconfig.base.json",
+        "ecmaFeatures": {
+            "jsx": true,
+        },
+    },
+    "plugins": [
+        "@typescript-eslint",
+    ],
+    "rules": {
```
Comment (LeonMontealegre) [inline linting/.core.js:22]: Maybe split this further into `.core.` and `.ts` for `@typescript-eslint specific` ones?

File: .eslintrc.js (hunk 2 above)
Comment (LeonMontealegre) [reply .eslintrc.js:78]: > So the whole .core file would be moved back into the main .eslintrc.js file?

No, not the .core `rules` or the TS `extends`, just the properties exclusive to the `.core` file like `env` and `ignorePatterns` and `parser` stuff

**Review by LeonMontealegre — APPROVED** [review]

**Pull request merged**